### ISSUE TITLE
GitLab Webhook Receives Bad Request

### ISSUE THREAD

**Issue by Dresdn** [mend/renovate-ce-ee#388]
Environment:
* GitLab SaaS
* Renovate-CE Docker v6.5.0

A few days ago, the GitLab webhook to Renovate started replying with just HTTP 400 "Bad Request". Bumping the log level to `debug` isn't showing anything, so I'm a bit at a loss on how to troubleshoot this.

I've tried clearing the /cache and /db to start fresh without luck. I know GitLab v16.5 was released a few weeks ago, but I'm unsure when my instance was updated.

Any ideas on how I can further troubleshoot this?

**Comment by Dresdn** [comment]
Want to add that the Renovate schedule isn't running either. The app is definitely authenticating to GitLab, but it's not doing anything. While in running in debug, these are the only the log messages:

```
DEBUG: Processing new task
       "task": {
         "id": 598,
         "reason": "hourly",
         "priority": 10,
         "action": "enqueue-all",
         "addedAt": "2023-11-09 22:00:00",
         "type": "internal"
       }
DEBUG: Task does not require model updating
       "task": {
         "id": 598,
         "reason": "hourly",
         "priority": 10,
         "action": "enqueue-all",
         "addedAt": "2023-11-09 22:00:00",
         "type": "internal"
       }
DEBUG: Done processing task
DEBUG: No new job to execute - waiting 15 seconds
DEBUG: No new job to execute - waiting 15 seconds
```

Seems like both directions of the GitLab integration aren't working.

**Comment by rarkins** [comment]
Please keep to one topic per discussion, and ideally one topic at a time

**Comment by rarkins** [comment]
@Gabriel-Ladzaretti I think we need to add a debug log for any 400 responses

**Comment by Gabriel-Ladzaretti** [comment]
A GitLab request will trigger a `400 Bad Response` (without a body) for the following reasons:

- Missing event type (warn)
- missing `x-gitlab-token` header (warn)
- Invalid `x-gitlab-token` (warn)
- Unsupported event (debug)
- An event without a repository name (info)

There's one exception: a missing log message when the `x-gitlab-event` header is absent.

We can make the logs clearer in these cases. Until then, I recommend reviewing the server logs for further investigation and, if possible, examining the GitLab request using the GitLab UI.

cc @PhilipAbed, @justo-mend.

**Comment by Dresdn** [comment]
@Gabriel-Ladzaretti - these are all being triggered directly from GitLab via the Test Webhooks action. Headers being sent are:

```
Content-Type: application/json
User-Agent: GitLab/16.6.0-pre
X-Gitlab-Event: Push Hook
X-Gitlab-Webhook-UUID: bd8b6647-3d5d-4dab-b5b2-7828d3f0ec54
X-Gitlab-Instance: https://gitlab.com
X-Gitlab-Token: [REDACTED]
X-Gitlab-Event-UUID: db214b62-257e-4249-94cb-13c00202f7f1
```

I also tried locally to ensure the `x-gitlab-token` is correct, and received the same response.

I'm using the Docker image, so the only logs I get are what node outputs.

**Comment by rarkins** [comment]
Let's answer with 200 instead of 400 though

**Comment by dennisameling** [comment]
I [switched](https://github.com/mend/renovate-ce-ee/issues/389#issuecomment-1820566576) to the new `MEND_RNV_AUTODISCOVER_FILTER`, and right after that, the webhooks started failing with `400 Bad Request`:

<img width="1050" alt="image" src="https://github.com/mend/renovate-ce-ee/assets/17739158/e1727017-9942-4ff7-9870-ffb894c2c8e5">

The `autodiscoverFilter` in the resolved configuration is looking good, and the Docker logs (even debug logs) don't show anything that would help me understand why the webhook is failing.

```
{
	"resolvedConfig": {
		"autodiscoverFilter": [
			"OUR_ORG_NAME/REPO_NAME",
			"OUR_ORG_NAME/REPO_NAME_2",
			".........."
		],
		"dataHandlerType": "sqlite",
		"sqliteVerbose": false,
		"workerHealthPort": "8080",
		"userAgent": "mend-renovate",
		"logHistoryCleanupCron": "0 0 * * *",
		"gitlabEndpoint": "https://gitlab.com/api/v4/",
		"serverPort": "8080",
		"platform": "gitlab",
		"gitlabPat": "***********",
		"licenseKey": "***********",
		"webhookBranchPrefixes": [
			"renovate/"
		],
		"sqliteFilePath": "/tmp/database.sqlite",
		"sqliteTraceMode": false,
		"workerExecutionTimeout": 3600000,
		"acceptTOS": "y",
		"serverApiRoot": "/api",
		"gitlabWebhookSecret": "***********",
		"cronAppSync": "0 0,4,8,12,16,20 * * *",
		"webhookBaseBranches": [
			"master",
			"main",
			"develop"
		],
		"adminApiEnabled": false,
		"logHistoryTTLDays": 30,
		"cronJobScheduler": "0 * * * *",
		"enqueueJobsOnStartup": "discovered"
	}
}
```

Whenever I remove the `MEND_RNV_AUTODISCOVER_FILTER` env variable, things immediately start working again. CC @Gabriel-Ladzaretti

**Comment by Gabriel-Ladzaretti** [comment]
Do these webhooks originate from filtered repositories? If they are not matched by the filter, the webhook will fail with 400 status code.

You can use the `api/job/add` [endpoint](https://github.com/mend/renovate-ce-ee/discussions/393#discussioncomment-7591553) to verify if the said repo is installed or not.
alternatively you can use the bundled sqlite3 cli tool and query the installed repos.

**Comment by Gabriel-Ladzaretti** [comment]
regardless, the 400 status code is indeed misleading and will be addressed.

**Comment by dennisameling** [comment]
> regardless, the 400 status code issue is indeed misleading (as it is expected) and will be addressed.

Ah, that's actually what's causing the problem here. When creating a group-wide webhook in GitLab, it fires a test request to validate that the webhook is working. But because Renovate now validates whether the request is coming from a configured repo, it fails with the 400 status code and GitLab disables the webhook, so not even the webhooks from the configured repos are coming in anymore.

I'll keep the filter disabled for now so that our webhook processing keeps working. Thanks for all your prompt replies, really appreciate it! 😊 

<img width="1078" alt="afbeelding" src="https://github.com/mend/renovate-ce-ee/assets/17739158/e55dcb8d-69c2-47e1-9bb4-c541db14c4f4">

**Comment by Dresdn** [comment]
I'm not experiencing that @dennisameling. We're also using Group Webhooks with `MEND_RNV_AUTODISCOVER_FILTER` configured. What I did notice is that the webhook was configured to send "Comments" too (maybe from a super old setup or from the Dependabot migration?), but the Group Hook test works just fine with just "Push Events" and "Merge Request Events" selected.

**Comment by dennisameling** [comment]
@Gabriel-Ladzaretti sorry for the ping, but has this been resolved in the meantime? Our Renovate instance is constantly at 80-100% CPU usage because of the large amount of (irrelevant) repos it has to go through. Thanks! 🙏🏼

**Comment by rarkins** [comment]
@dennisameling my understanding of the original request is simply return a 200 instead of 400 when it doesn't match a filtered repo. But how would that solve any CPU problem?

**Comment by rarkins** [comment]
And BTW it's expected that a Renovate CE instance is at high CPU if it's running repo jobs. The only time it won't be at high CPU is if it's idle (e.g. it finishes every job)

**Comment by dennisameling** [comment]
> @dennisameling my understanding of the original request is simply return a 200 instead of 400 when it doesn't match a filtered repo.

Correct.

> But how would that solve any CPU problem?

Not directly of course, but because we had to disable `MEND_RNV_AUTODISCOVER_FILTER` entirely because it causes this issue with the webhooks returning 400 (and therefore GitLab entirely disabling the webook automatically), Renovate now scans _all_ our repos, even if it's not configured there (it creates onboarding PRs all over the place). We now have a permanent job queue of 100+ repos which we wouldn't have if we could filter out many irrelevant repos with `MEND_RNV_AUTODISCOVER_FILTER`. Before this functionality broke, our ECS instance had more than enough capacity to run Renovate on all the repos that _are_ relevant to us.

Here you can see what happened after we upgraded Renovate back in November and had to disable the autodiscover filter:

<img width="449" alt="afbeelding" src="https://github.com/mend/renovate-ce-ee/assets/17739158/ec3c370e-fa21-4997-b764-0febe907d507">

**Comment by Gabriel-Ladzaretti** [comment]
> @dennisameling my understanding of the original request is simply return a 200 instead of 400 when it doesn't match a filtered repo. But how would that solve any CPU problem?

This was addressed in `v6.7.0`, which was released on November 27, 2023. We no longer response with `400` for bad webhook requests.

**Comment by dennisameling** [comment]
I can confirm the `MEND_RNV_AUTODISCOVER_FILTER` is working as expected now. We just weren't aware because this issue was still open. Thanks very much 😊  🎉
